### PR TITLE
[Merged by Bors] - Add ParallelCommands system parameter

### PR DESCRIPTION
**Pull request by TheRawMeatball**
(follow-up to #4423)
# Objective
Currently, it isn't possible to easily fire commands from within par_for_each blocks. This PR allows for issuing commands from within parallel scopes.

### CODE REVIEW

File: crates/bevy_ecs/src/system/commands/mod.rs [hunk 1]
```diff
@@ -55,6 +57,11 @@ impl<'w, 's> Commands<'w, 's> {
         }
     }
 
+    /// Create a new `Commands` from a queue and an [`Entities`] reference.
+    pub fn new_from_entities(queue: &'s mut CommandQueue, entities: &'w Entities) -> Self {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/commands/mod.rs:61]: Can we get a quick doc example of how you'd actually access this in e.g. a test?

File: crates/bevy_ecs/src/system/commands/parallel_scope.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+use crate::{
+    entity::Entities,
+    prelude::World,
+    system::{SystemParam, SystemParamFetch, SystemParamState},
+};
+
+use super::{CommandQueue, Commands};
+
+#[derive(Default)]
+pub struct ParallelCommandsState {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/commands/parallel_scope.rs:14]:
```suggestion
/// The internal state of a [`ParallelCommands`] object
pub struct ParallelCommandsState {
```

Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/commands/parallel_scope.rs:14]:
```suggestion
struct ParallelCommandsState {
```

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/commands/parallel_scope.rs:14]: Unfortunately, it appears that this must be pub, because it's the associated type for a public trait.

**Comment by alice-i-cecile** [comment]
Very nice. I'm looking forward to https://github.com/bevyengine/rfcs/pull/54 so we can clean this up even further.

File: crates/bevy_ecs/Cargo.toml [hunk 3]
```diff
@@ -21,6 +21,7 @@ bevy_utils = { path = "../bevy_utils", version = "0.8.0-dev" }
 bevy_ecs_macros = { path = "macros", version = "0.8.0-dev" }
 
 async-channel = "1.4"
+thread_local = "1.1.4"
```
Comment (DJMcNab) [inline crates/bevy_ecs/Cargo.toml:24]: Is this already in our tree?

Comment (TheRawMeatball) [reply crates/bevy_ecs/Cargo.toml:24]: No, but it is a nice crate that another PR (#4663) depends on as well.

File: crates/bevy_ecs/src/system/commands/mod.rs (hunk 1 above)
Comment (TheRawMeatball) [reply crates/bevy_ecs/src/system/commands/mod.rs:61]: Unfortunately, it seems that so far this method cannot be effectively used, as `SystemParam` derives cannot currently implement custom apply strategies, and I don't think it'd be a great idea to add an entire manual `SystemParam` impl here to use as an example. I'll make a separate PR for custom apply strategies in `SystemParam` derives later on, which could then be used as an example here.

File: crates/bevy_ecs/src/system/commands/parallel_scope.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+use crate::{
+    entity::Entities,
+    prelude::World,
+    system::{SystemParam, SystemParamFetch, SystemParamState},
+};
+
+use super::{CommandQueue, Commands};
+
+#[derive(Default)]
```
Comment (james7132) [inline crates/bevy_ecs/src/system/commands/parallel_scope.rs:13]:
```suggestion
#[doc(hidden)]
#[derive(Default)]
```

File: crates/bevy_ecs/src/system/commands/parallel_scope.rs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    fn apply(&mut self, world: &mut World) {
+        for cq in self.tls.iter_mut() {
+            cq.get_mut().apply(world);
+        }
+    }
+}
+
+impl<'w, 's> ParallelCommands<'w, 's> {
+    pub fn command_scope<R>(&self, f: impl FnOnce(Commands) -> R) -> R {
+        let tls = &self.state.tls;
```
Comment (james7132) [inline crates/bevy_ecs/src/system/commands/parallel_scope.rs:85]: nit: Prefer more self-describing variable names, two/three character names are still a bit hard to follow.

**Review by james7132 — CHANGES_REQUESTED** [review]

File: crates/bevy_ecs/src/system/commands/parallel_scope.rs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+use super::{CommandQueue, Commands};
+
+#[derive(Default)]
+/// The internal [`SystemParamState`] of the [`ParallelCommands`] type
+pub struct ParallelCommandsState {
+    tls: ThreadLocal<Cell<CommandQueue>>,
+}
+
+/// An alternative to [`Commands`] that can be used in parallel contexts, such as those in [`Query::par_for_each`]
+///
```
Comment (james7132) [inline crates/bevy_ecs/src/system/commands/parallel_scope.rs:20]: One more thought: probably should leave a note here saying that command application order is dependent on the number of threads, so any non-idempotent commands may result in different results on different machines.

**Review by james7132 — CHANGES_REQUESTED** [review]

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ecs/src/system/commands/parallel_scope.rs (hunk 6 above)
Comment (TheRawMeatball) [reply crates/bevy_ecs/src/system/commands/parallel_scope.rs:21]: Did you mean "non-commutative"? Given that commands take `self` and can't be executed twice idempotence is a bit of a moot point.

Comment (james7132) [reply crates/bevy_ecs/src/system/commands/parallel_scope.rs:23]: Sorry, was thinking of this through the lens of networking where command application like this is a potential concern for determinism. Yes you're right.

**Review by james7132 — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
I've merged in #4705; @TheRawMeatball can you rebase to update the doc example with that change?

Once that's done I'll merge this in.

File: crates/bevy_ecs/src/system/commands/parallel_scope.rs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    system::{SystemParam, SystemParamFetch, SystemParamState},
+};
+
+use super::{CommandQueue, Commands};
+
+#[doc(hidden)]
+#[derive(Default)]
+/// The internal [`SystemParamState`] of the [`ParallelCommands`] type
+pub struct ParallelCommandsState {
+    thread_local_storage: ThreadLocal<Cell<CommandQueue>>,
```
Comment (james7132) [inline crates/bevy_ecs/src/system/commands/parallel_scope.rs:17]: I wonder if we should generalize this and add it to `bevy_utils`. This is seemingly a very useful abstraction that showed up in #4899 too.

**Review by alice-i-cecile — APPROVED** [review]
